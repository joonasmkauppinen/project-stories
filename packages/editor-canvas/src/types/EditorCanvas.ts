export type EntityState = 'idle' | 'hovered' | 'active';

export type LayerAnimationType =
  | 'drop'
  | 'fade-in'
  | 'fly-in-bottom'
  | 'fly-in-left'
  | 'fly-in-right'
  | 'fly-in-top'
  | 'pan-down'
  | 'pan-left'
  | 'pan-right'
  | 'pan-up'
  | 'pulse'
  | 'rotate-in-left'
  | 'rotate-in-right'
  | 'twirl-in'
  | 'whoosh-in-left'
  | 'whoosh-in-right'
  | 'zoom-in'
  | 'zoom-out';

export type MimeType =
  | 'audio/mpeg'
  | 'image/gif'
  | 'image/jpeg'
  | 'image/png'
  | 'image/svg+xml'
  | 'text/plain'
  | 'video/mp4';

export type LayerType = 'text' | 'image' | 'video' | 'audio' | 'svg';

export interface LayerCoordinate {
  x: number;
  y: number;
}

export interface LayerResource {
  alt: string;
  height: number;
  id: string;
  mimeType: MimeType;
  src: string;
  type: LayerType;
  width: number;
}

// export interface TextResource {
//   id: string;
//   value: string;
//   mimeType: 'text/plain';
// }

// export interface VideoResource {
//   id:
// }

export interface CardLayer {
  animation?: {
    type: LayerAnimationType;
    delay?: number;
    duration?: number;
  };
  id: string;
  // position: LayerCoordinate;
  // resource: LayerResource;
  // state: EntityState;
  type: LayerType;
}

export interface Card {
  autoAdvance: boolean;
  duration: number;
  id: string;
  layers: CardLayer[];
  name: string;
  state: EntityState;
}

export type OnHoverCallback = (props: {
  parentId?: string;
  id: string;
}) => void;
