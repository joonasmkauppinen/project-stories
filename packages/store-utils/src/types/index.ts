export * from './storeActions/storeActions';
export * from './appState/appState';
