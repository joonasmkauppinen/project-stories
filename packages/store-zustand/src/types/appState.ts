export type CardState = 'idle' | 'hovered' | 'active';

export type LayerState = 'idle' | 'hovered' | 'active';

export interface LayerMetaState {
  parentCardActive: boolean;
}

export type ID = string;

export type LayerAnimationType =
  | 'drop'
  | 'fade-in'
  | 'fly-in-bottom'
  | 'fly-in-left'
  | 'fly-in-right'
  | 'fly-in-top'
  | 'pan-down'
  | 'pan-left'
  | 'pan-right'
  | 'pan-up'
  | 'pulse'
  | 'rotate-in-left'
  | 'rotate-in-right'
  | 'twirl-in'
  | 'whoosh-in-left'
  | 'whoosh-in-right'
  | 'zoom-in'
  | 'zoom-out';

export type MimeType =
  | 'audio/mpeg'
  | 'image/gif'
  | 'image/jpeg'
  | 'image/png'
  | 'image/svg+xml'
  | 'text/plain'
  | 'video/mp4';

export type LayerType = 'text' | 'image' | 'video' | 'audio' | 'svg';

export interface Coordinate {
  x: number;
  y: number;
}

export interface Size {
  width: number;
  height: number;
}

export interface LayerResource {
  alt: string;
  height: number;
  id: ID;
  mimeType: MimeType;
  src: string;
  type: LayerType;
  width: number;
}

export interface BaseLayer {
  animation?: {
    type: LayerAnimationType;
    delay?: number;
    duration?: number;
  };
  sortOrderIndex: number;
  position: Coordinate;
  screenPosition: Coordinate;
  size: Size;
  state: LayerState;
  metaState: LayerMetaState;
  name: string;
  type?: LayerType;
}

export interface TextLayer extends BaseLayer {
  type: 'text';
  value: string;
}

export type Layer = TextLayer;

export interface Layers {
  [key: string]: Layer;
}

export interface Card {
  sortOrderIndex: number;
  screenPosition: Coordinate;
  autoAdvance: boolean;
  duration: number;
  layers: Layers;
  name: string;
  state: CardState;
}

export interface Cards {
  [key: string]: Card;
}

// TODO: refactor to SelectedLayers and SelectedCards
export interface SelectionItem {
  id: ID;
  parentId?: ID;
}

export interface SelectedLayer {
  cardId: ID;
  layerId: ID;
}

export interface SelectedCard {
  cardId: ID;
}

export type StoryboardTool = 'move' | 'text' | 'hand';

export interface UserInteraction {
  isDragging: boolean;
  isEditingText: boolean;
}

export interface AppState {
  cards: Cards;
  currentTool: StoryboardTool;
  selectedLayers: SelectedLayer[];
  selectedCards: SelectedCard[];
  userInteraction: UserInteraction;
}
