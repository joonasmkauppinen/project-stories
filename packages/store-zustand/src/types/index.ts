export * from './appState';
export * from './layerActions';
export * from './testUtils';
