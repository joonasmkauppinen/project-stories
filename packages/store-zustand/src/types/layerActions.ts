import {
  AddNewCard,
  AddTextLayerToCard,
  DeleteSelectedLayers,
  DeselectAll,
  OnDragSelection,
  SetCardStateToActive,
  SetCardStateToHovered,
  SetCardStateToIdle,
  SetIsDraggingToFalse,
  SetIsDraggingToTrue,
  SetLayerStateToActive,
  SetLayerStateToHovered,
  SetLayerStateToIdle,
  SetToolToMove,
  SetToolToText,
  UpdateElementScreenPosition,
} from '../actions';

export interface LayerActions {
  addNewCard: AddNewCard;
  addTextLayerToCard: AddTextLayerToCard;
  deleteSelectedLayers: DeleteSelectedLayers;
  deselectAll: DeselectAll;
  onDragSelection: OnDragSelection;
  setCardStateToActive: SetCardStateToActive;
  setCardStateToHovered: SetCardStateToHovered;
  setCardStateToIdle: SetCardStateToIdle;
  setIsDraggingToFalse: SetIsDraggingToFalse;
  setIsDraggingToTrue: SetIsDraggingToTrue;
  setLayerStateToActive: SetLayerStateToActive;
  setLayerStateToHovered: SetLayerStateToHovered;
  setLayerStateToIdle: SetLayerStateToIdle;
  setToolToMove: SetToolToMove;
  setToolToText: SetToolToText;
  updateElementScreenPosition: UpdateElementScreenPosition;
}

export interface LayerActionsProp {
  actions: LayerActions;
}
