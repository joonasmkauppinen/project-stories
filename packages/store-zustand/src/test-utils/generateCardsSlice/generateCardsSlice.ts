import { generateCard, generateTextLayer } from '../../generators';
import { Card, Layer, TestCardId, TestLayerId } from '../../types';

interface TestLayer extends Partial<Omit<Layer, 'sortOrderIndex'>> {
  id: TestLayerId;
}

interface TestCard extends Partial<Omit<Card, 'layers' | 'sortOrderIndex'>> {
  id: TestCardId;
  layers?: TestLayer[];
}

export const generateCardsSlice = (cards: TestCard[]) => {
  return Object.fromEntries(
    cards.map((card, index) => {
      const { layers, id, ...rest } = card;

      const generateLayers = (layers: TestLayer[]) =>
        Object.fromEntries(
          layers.map((layer, index) => {
            const { id, ...rest } = layer;
            return generateTextLayer({
              sortOrderIndex: index,
              top: 50 * index,
              type: 'text',
              testOverrides: { id: id, properties: rest },
            }).idWithData;
          })
        );

      return generateCard({
        sortOrderIndex: index,
        testOverrides: { id: id, properties: rest },
        layers: layers ? generateLayers(layers) : {},
      }).idWithData;
    })
  );
};
