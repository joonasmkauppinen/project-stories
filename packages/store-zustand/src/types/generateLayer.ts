import { ID } from './appState';

export interface GenerateLayerReturnType<LayerType> {
  layerId: ID;
  layerData: LayerType;
  idWithData: [string, LayerType];
}
