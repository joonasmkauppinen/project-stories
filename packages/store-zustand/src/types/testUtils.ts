import { Card, Layer } from './appState';

export type TestCardId =
  | 'test_card_id_0'
  | 'test_card_id_1'
  | 'test_card_id_2'
  | 'test_card_id_3';

export interface TestCardOverrides {
  id?: TestCardId;
  properties?: Partial<Omit<Card, 'sortOrderIndex'>>;
}

export type TestLayerId =
  | 'test_layer_id_0'
  | 'test_layer_id_1'
  | 'test_layer_id_2'
  | 'test_layer_id_3';

export interface TestLayerOverrides {
  id?: TestLayerId;
  properties?: Partial<Omit<Layer, 'sortOrderIndex'>>;
}
