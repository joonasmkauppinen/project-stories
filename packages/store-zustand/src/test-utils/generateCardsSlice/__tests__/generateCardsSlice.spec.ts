import { TestCardId, TestLayerId } from '../../../types';

import { generateCardsSlice } from '../generateCardsSlice';

describe('test-util - generateCardsSlice()', () => {
  test('Generates one card with no layers correctly', () => {
    const oneCardWithoutLayers = generateCardsSlice([
      {
        id: 'test_card_id_0',
      },
    ]);

    expect(oneCardWithoutLayers).toMatchSnapshot();
  });

  test('Generates one card with layers correctly', () => {
    const oneCardWithLayers = generateCardsSlice([
      {
        id: 'test_card_id_0',
        layers: [{ id: 'test_layer_id_0' }],
      },
    ]);

    expect(oneCardWithLayers).toMatchSnapshot();
  });

  test('Generates two cards with no layers correctly', () => {
    const twoCardsWithoutLayers = generateCardsSlice([
      {
        id: 'test_card_id_0',
      },
      {
        id: 'test_card_id_1',
      },
    ]);

    expect(twoCardsWithoutLayers).toMatchSnapshot();
  });

  test('Generates two cards with layers correctly', () => {
    const twoCardsWithLayers = generateCardsSlice([
      {
        id: 'test_card_id_0',
        layers: [{ id: 'test_layer_id_0' }],
      },
      {
        id: 'test_card_id_1',
        layers: [{ id: 'test_layer_id_0' }, { id: 'test_layer_id_1' }],
      },
    ]);

    expect(twoCardsWithLayers).toMatchSnapshot();
  });

  test('Sets card "state" override property correctly', () => {
    const cardId: TestCardId = 'test_card_id_0';
    const cardWithStateActive = generateCardsSlice([
      { id: cardId, state: 'active' },
    ]);

    expect(cardWithStateActive[cardId].state).toBe('active');
  });

  test('Sets layer "state" override property correctly', () => {
    const cardId: TestCardId = 'test_card_id_0';
    const layerId: TestLayerId = 'test_layer_id_0';
    const cardWithLayerStateActive = generateCardsSlice([
      { id: cardId, layers: [{ id: layerId, state: 'active' }] },
    ]);

    expect(cardWithLayerStateActive[cardId].layers[layerId].state).toBe(
      'active'
    );
  });
});
