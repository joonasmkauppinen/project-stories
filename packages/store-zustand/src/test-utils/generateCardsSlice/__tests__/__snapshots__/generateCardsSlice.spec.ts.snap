// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-util - generateCardsSlice() Generates one card with layers correctly 1`] = `
Object {
  "test_card_id_0": Object {
    "autoAdvance": false,
    "duration": 0,
    "layers": Object {
      "test_layer_id_0": Object {
        "metaState": Object {
          "parentCardActive": false,
        },
        "name": "text 1",
        "position": Object {
          "x": 20,
          "y": 0,
        },
        "screenPosition": Object {
          "x": NaN,
          "y": NaN,
        },
        "size": Object {
          "height": NaN,
          "width": 320,
        },
        "sortOrderIndex": 0,
        "state": "idle",
        "type": "text",
        "value": "Hello World",
      },
    },
    "name": "Card 1",
    "screenPosition": Object {
      "x": NaN,
      "y": NaN,
    },
    "sortOrderIndex": 0,
    "state": "idle",
  },
}
`;

exports[`test-util - generateCardsSlice() Generates one card with no layers correctly 1`] = `
Object {
  "test_card_id_0": Object {
    "autoAdvance": false,
    "duration": 0,
    "layers": Object {},
    "name": "Card 1",
    "screenPosition": Object {
      "x": NaN,
      "y": NaN,
    },
    "sortOrderIndex": 0,
    "state": "idle",
  },
}
`;

exports[`test-util - generateCardsSlice() Generates two cards with layers correctly 1`] = `
Object {
  "test_card_id_0": Object {
    "autoAdvance": false,
    "duration": 0,
    "layers": Object {
      "test_layer_id_0": Object {
        "metaState": Object {
          "parentCardActive": false,
        },
        "name": "text 1",
        "position": Object {
          "x": 20,
          "y": 0,
        },
        "screenPosition": Object {
          "x": NaN,
          "y": NaN,
        },
        "size": Object {
          "height": NaN,
          "width": 320,
        },
        "sortOrderIndex": 0,
        "state": "idle",
        "type": "text",
        "value": "Hello World",
      },
    },
    "name": "Card 1",
    "screenPosition": Object {
      "x": NaN,
      "y": NaN,
    },
    "sortOrderIndex": 0,
    "state": "idle",
  },
  "test_card_id_1": Object {
    "autoAdvance": false,
    "duration": 0,
    "layers": Object {
      "test_layer_id_0": Object {
        "metaState": Object {
          "parentCardActive": false,
        },
        "name": "text 1",
        "position": Object {
          "x": 20,
          "y": 0,
        },
        "screenPosition": Object {
          "x": NaN,
          "y": NaN,
        },
        "size": Object {
          "height": NaN,
          "width": 320,
        },
        "sortOrderIndex": 0,
        "state": "idle",
        "type": "text",
        "value": "Hello World",
      },
      "test_layer_id_1": Object {
        "metaState": Object {
          "parentCardActive": false,
        },
        "name": "text 2",
        "position": Object {
          "x": 20,
          "y": 50,
        },
        "screenPosition": Object {
          "x": NaN,
          "y": NaN,
        },
        "size": Object {
          "height": NaN,
          "width": 320,
        },
        "sortOrderIndex": 1,
        "state": "idle",
        "type": "text",
        "value": "Hello World",
      },
    },
    "name": "Card 2",
    "screenPosition": Object {
      "x": NaN,
      "y": NaN,
    },
    "sortOrderIndex": 1,
    "state": "idle",
  },
}
`;

exports[`test-util - generateCardsSlice() Generates two cards with no layers correctly 1`] = `
Object {
  "test_card_id_0": Object {
    "autoAdvance": false,
    "duration": 0,
    "layers": Object {},
    "name": "Card 1",
    "screenPosition": Object {
      "x": NaN,
      "y": NaN,
    },
    "sortOrderIndex": 0,
    "state": "idle",
  },
  "test_card_id_1": Object {
    "autoAdvance": false,
    "duration": 0,
    "layers": Object {},
    "name": "Card 2",
    "screenPosition": Object {
      "x": NaN,
      "y": NaN,
    },
    "sortOrderIndex": 1,
    "state": "idle",
  },
}
`;
